src/init/starship.zsh

```diff
@@ -41,23 +41,23 @@ starship_precmd() {
     # quotes so we set it here and then use the value later on.
     STARSHIP_JOBS_COUNT=${#jobstates}
 }
-starship_preexec() {
+prompt_starship_preexec() {
     __starship_get_time && STARSHIP_START_TIME=$STARSHIP_CAPTURED_TIME
 }
 
 # If precmd/preexec arrays are not already set, set them. If we don't do this,
-# the code to detect whether starship_precmd is already in precmd_functions will
-# fail because the array doesn't exist (and same for starship_preexec)
+# the code to detect whether prompt_starship_precmd is already in precmd_functions will
+# fail because the array doesn't exist (and same for prompt_starship_preexec)
 (( ! ${+precmd_functions} )) && precmd_functions=()
 (( ! ${+preexec_functions} )) && preexec_functions=()
 
 # If starship precmd/preexec functions are already hooked, don't double-hook them
 # to avoid unnecessary performance degradation in nested shells
-if [[ -z ${precmd_functions[(re)starship_precmd]} ]]; then
-    precmd_functions+=(starship_precmd)
+if [[ -z ${precmd_functions[(re)prompt_starship_precmd]} ]]; then
+    precmd_functions+=(prompt_starship_precmd)
```

Comment:
Maybe it can use [`add-zsh-hook` module](https://zsh.sourceforge.io/Doc/Release/User-Contributions.html#Manipulating-Hook-Functions) of zsh, this method is also recommended by [zsh official theme document](https://zsh.sourceforge.io/Doc/Release/User-Contributions.html#Writing-Themes). 
Using this method is good for implementation of https://github.com/starship/starship/issues/1994 as well.

Reply:
Thanks for the catch.

One thing is that right now, the ZSH init of starship is written so that you can run `eval $(starship init zsh)` inside a clean copy of ZSH with no modules loaded (e.g. one started with `zsh -f`) and it will work.

Using `add-zsh-hook` breaks this because you then then need to `autoload` some stuff, and I'm not sure how comfortable I am `autoloading` things on behalf of the user (e.g. if we try to load a module that isn't compiled in their specific version of ZSH, the entire init becomes unusable).

I think I would prefer to use this as a springboard to create another init function (e.g. `starship init zsh --gen-zsh-prompt [path]` or something) that creates the ZSH-style prompt files. However, if we can show that the add-zsh-hook function is always autoloadable, even when e.g. ZSH was compiled with modules and the modules are now missing, then maybe we could autoload it for the user in the main init script.

Reply:
On the flip side of this, we already do some `zmodload`ing to access the job states, so I'm not really sure if add-zsh-hook is really that much more of an issue....

Reply:
It’s fairly normal for 3rd-party Zsh scripts to `autoload` functions or `zmodload` modules that ship with Zsh itself. I wouldn’t worry about it.

Reply:
@CoelacanthusHex Can the conversion to `add-zsh-hook` be done as a separate MR, please? Let’s not bloat the scope of this one needlessly.

Reply:
> Can the conversion to add-zsh-hook be done as a separate MR, please? Let’s not bloat the scope of this one needlessly.

Ok, I will create a PR to do it after this pr merged, because of function name changes.

Reply:
> @CoelacanthusHex Can the conversion to `add-zsh-hook` be done as a separate MR, please? Let’s not bloat the scope of this one needlessly.

Oh, because this PR delayed for a week, so I create PR before this PR merged. 
https://github.com/starship/starship/pull/3537
If my pr merged first, this pr need rebase to resolve conflict, vice versa.